career-mentorship-program/2021-spring/matching_code.py

```diff
@@ -373,28 +456,38 @@ def compute_weight_and_statistics(self):
 
         CAREER_INTERESTS_MULTIPLIER = 5
         RECRUITMENT_TOPICS_MULTIPLIER = 3
-        MENTOR_NOT_OLDER_PENALTY = 10000000 # requires the mentor to be older
+        HOBBIES_TOPICS_MULTIPLIER = 1.5
+        MENTOR_NOT_OLDER_PENALTY = 10000000  # requires the mentor to be older
+        MENTOR_LESS_CS_EXPERIENCE_PENALTY = 10000000  # requires the mentor to be more experienced
         MENTOR_CLOSE_IN_YEAR_BONUS = 1
         MENTOR_TAKES_ON_ANOTHER_MENTEE_PENALTY = 100
+        GENDER_SAME_BONUS = 20  # bonus if both prefer same gender and are same gender
+        RACE_SAME_BONUS = 20  # bonus if both prefer same race and are same race
+        TIME_ZONE_DIFFERENCE_PENALTY = 5
+        TIME_AVAILABLE_DIFFERENCE_PENALTY = 20
 
         ### Compute weight based on matching critera ###
 
         weight = 0
         statistics = {}
 
+        # function to compute topic overlaps
```

Comment:
In general, I find it best to not put side effects (like altering the weight), inside a function like this, because it can be confusing to the reader where the side effect is happening. I personally would find it easier to read if all of the weight-editing logic were in the main function. What I would recommend having this function do is simply take in two sets and compute the fraction of overlap and return it. This is what I was attempting to convey during our code walk through.

```
# Note that the order in which you pass things in here matters
def compute_fraction_overlap(mentee_set, mentor_set):
    return len(mentee_set & mentor_set) / len(mentee_set)
```
   
Then in the larger function you can do:

```
interests_overlap_fraction = compute_fraction_overlap(self.mentee.interested_fields, self.mentor.experienced_fields)
statistics["interests_overlap_fraction"] = interests_overlap_fraction
weight += CAREER_INTERESTS_MULTIPLIER * interests_overlap_fraction
```